src/platforms/esp32/components/avm_builtins/storage_nif.c

```diff
@@ -0,0 +1,299 @@
+    int part_by_name_len = strlen(part_by_name_prefix);
+    if (!strncmp(part_by_name_prefix, source, part_by_name_len)) {
+        mount_config.allocation_unit_size = CONFIG_WL_SECTOR_SIZE;
+
+        struct MountedFS *mount = malloc(sizeof(struct MountedFS));
+        mount->misc_entry.entry_type = (uintptr_t) nif_esp_mount;
+        mount->base_path = strdup(target);
+        mount->mount_type = FATSPIFlash;
+        synclist_append(&platform->misc_entries, &mount->misc_entry.list_head);
+
+#if ESP_IDF_VERSION_MAJOR >= 5
+        ret = esp_vfs_fat_spiflash_mount_rw_wl(
+            target, source + part_by_name_len, &mount_config, &mount->handle.wl);
+#else
+        ret = esp_vfs_fat_spiflash_mount(
+            target, source + part_by_name_len, &mount_config, &mount->handle.wl);
+#endif
+
+// C3 doesn't support this
+#ifdef SDMMC_SLOT_CONFIG_DEFAULT
+    } else if (!strcmp(source, "sdmmc")) {
+        mount_config.allocation_unit_size = 512;
+
+        sdmmc_host_t host_config = SDMMC_HOST_DEFAULT();
+        sdmmc_slot_config_t slot_config = SDMMC_SLOT_CONFIG_DEFAULT();
+
+        struct MountedFS *mount = malloc(sizeof(struct MountedFS));
+        mount->misc_entry.entry_type = (uintptr_t) nif_esp_mount;
+        mount->base_path = strdup(target);
+        mount->mount_type = FATSDMMC;
+        synclist_append(&platform->misc_entries, &mount->misc_entry.list_head);
+
+        ret = esp_vfs_fat_sdmmc_mount(
+            target, &host_config, &slot_config, &mount_config, &mount->handle.card);
+#endif
+
+    } else if (!strcmp(source, "sdspi")) {
+        mount_config.allocation_unit_size = 512;
+
+        sdmmc_host_t host_config = SDSPI_HOST_DEFAULT();
+        sdspi_device_config_t spi_slot_config = SDSPI_DEVICE_CONFIG_DEFAULT();
+
+        term spi_port = interop_kv_get_value_default(
+            opts_term, ATOM_STR("\x8", "spi_host"), term_invalid_term(), ctx->global);
+        spi_host_device_t host_dev;
+        bool ok = spi_driver_get_peripheral(spi_port, &host_dev, ctx->global);
+        if (!ok) {
+            RAISE_ERROR(BADARG_ATOM);
+        }
+        spi_slot_config.host_id = host_dev;
+
+        term cs_term = interop_kv_get_value_default(
+            opts_term, ATOM_STR("\x2", "cs"), term_invalid_term(), ctx->global);
+        VALIDATE_VALUE(cs_term, term_is_integer);
+        spi_slot_config.gpio_cs = term_to_int(cs_term);
+
+        term cd_term = interop_kv_get_value_default(
+            opts_term, ATOM_STR("\x2", "cd"), UNDEFINED_ATOM, ctx->global);
+        if (cd_term != UNDEFINED_ATOM) {
+            VALIDATE_VALUE(cd_term, term_is_integer);
+            spi_slot_config.gpio_cd = term_to_int(cd_term);
+        }
+
+        struct MountedFS *mount = malloc(sizeof(struct MountedFS));
+        mount->misc_entry.entry_type = (uintptr_t) nif_esp_mount;
+        mount->base_path = strdup(target);
+        mount->mount_type = FATSPIFlash;
+        synclist_append(&platform->misc_entries, &mount->misc_entry.list_head);
```

Comment:
Instead of altering platform data, why don't you use a resource? It would unmount if garbage collected, and would make iterating on the list of mount points unnecessary.

Reply:
That was one of my thoughts, but I wanted to mimic the same semantic that exists on regular unix where disks are unmounted using the path as handle, so in case we generalize our API we are already support the same semantics.

But, after your feedback, I was thinking that one day we can store paths inside of refs on systems that uses paths as handles for unmounting.

Does it makes sense to you?

Reply:
Definitely. If the system requires paths to unmount, it is unlikely very short on memory so we can maintain a copy around of the path in the resource. While on ESP32, a resource will consume a little less than the linked list with the copy of the path.

Reply:
@pguyot on a 3rd thought there are other problems with this approach:
- Basically open files are resources, so we need to make sure that they are closed and garbage collected before running umount.
- I dind't check if umounting can cause errors when files are open, but I don't think there is an option to postpone destructor and running it again when all files have been closed, in order to try umount a second time
- I'm not sure if performing I/O (that might also fail) from a resource destructor is a good idea, but here I can change my mind

Reply:
I never thought umount should only be implicit. It should be explicit and resource garbage collection should serve as a fallback. The same thing is true with files. Close should be explicitely called.

If necessary, we can increase the ref count of resources that need to be destroyed later. But in this case, it shouldn't be the case. It's a programming error to not explicitely call umount or close, but an error that doesn't leak with a resource (and does without a resource).

Reply:
I think I fixed this comment.